File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        maxC = 0
        l, r = 0, 1
        
        while r < len(prices) and l < r:
            if prices[r] < prices[l]:
                l = r
                r = r + 1
                continue
            else:
                maxC = max(maxC, prices[r] - prices[l])
                r += 1
        return maxC

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.c
int maxProfit(int* prices, int pricesSize){
    int* dp = (int*)malloc(sizeof(int) * pricesSize);
    for(int i=0; i < pricesSize; i++){
        dp[i] = 0;
    }
    int m = prices[pricesSize - 1];
    for(int i=pricesSize - 2; i >= 0; i-- ){
        dp[i] = dp[i + 1] >= m - prices[i] ? dp[i + 1] : m - prices[i];
        m = prices[i] >= m ? prices[i] : m;
    };

    int final = dp[0];
    free(dp);
    return final;
}